### ISSUE TITLE
Is the cubemap example correct? (Flipped church)

### ISSUE THREAD

**Issue by mkvenkit** [webgpu/webgpu-samples#341]
Thanks for the cubemap example!

But I am a bit confused about the position of the church. When I arrange the cubemap images in the standard "+" format, I get the first image below. You can see in the -Z tile that the church is to the **left** of the bridge. Now look at the second image, which shows your example. The church is on the **right** of the bridge. Is the mapping correct? Currently the WebGPU documentation does not document the "cube" texture convention in detail. Does it match the OpenGL cubemap conventions in the link below? Thanks for any clarification!

https://www.khronos.org/opengl/wiki/Cubemap_Texture

![bridge](https://github.com/webgpu/webgpu-samples/assets/35648873/c6f52320-279a-4df8-ad4a-17d57aa3db23)

<img width="940" alt="Screenshot 2023-12-20 at 9 55 22 PM" src="https://github.com/webgpu/webgpu-samples/assets/35648873/5f636e33-722a-490d-9f40-421e92211fb0">

**Comment by greggman** [comment]
I think you are right.

WebGPU and WebGL are the same here. Using these textures the WebGL example on [this page](https://webgpufundamentals.org/webgpu/lessons/webgpu-skybox.html) shows [the same issue](https://jsgist.org/?src=552873ee6d90ccf49c2031a8155bebe0). And the WebGPU example on [this page](https://webgpufundamentals.org/webgpu/lessons/webgpu-skybox.html) also shows [the same issue](https://jsgist.org/?src=be5d1eb4487de4fdcbd6e4d2de3aa087)

I haven't really thought through what the "correct" solution is. Maybe it's the fact that you're inside the cube but a cube map is defined from outside?

For any of the examples, a simple solution is to multiply the the normal at texture lookup time by [-1, 1, 1]

       return textureSample(ourTexture, ourSampler, normalize(t.xyz / t.w) * vec3f(-1, 1, 1));

**Comment by mkvenkit** [comment]
Thanks for the quick response, @greggman. 

My understanding is that in both the real world and the virual world a cube map is constructed by capturing a picture/render in the 6 directions. This is done from inside the cube, so the image tiles in a cube map are the "inside view". Also, as per the GL docs, cube map uses as left-handed coordinate system. The WGSL spec says:

```
@must_use fn textureSample(t: T,
                           s: sampler,
                           coords: vec3<f32>) -> vec4<f32> 

T is texture_3d<f32>, or texture_cube<f32>	                           
```

So what's the coordinate system of `coords`? If it is expecting a left handed system, then we need to apply a scale of `S(1, 1, -1)` to our world coordinate `coords`. I think. It would be good for the WebGPU spec to clarify this in the documentation.

In any case, I'll try some experiments.

**Comment by mkvenkit** [comment]
Hi @greggman, the scale I mentioned above seems to fix the problem. I have a live demo here, which also implements environment mapping:

https://electronut.in/webgpu/ch4_texturing/cubemap/

There are 2 photo features you can check for symmetry. The dark/light vases in front of buddha and the left/right boxes near the entrance  - the left one is ornamented. 

Thanks!

**Comment by greggman** [comment]
I'm not sure it matters which you flip x or z. But, if you take your example and move the camera to 0, 5, 50 then according to your code the camera is at 0, 5, 50, looking at 0, 10, 0 which means it's looking at negative-z

<img width="1390" alt="Screenshot 2023-12-21 at 01 36 30" src="https://github.com/webgpu/webgpu-samples/assets/234804/bab7b015-33c5-48fa-a390-d96c4bf17473">

But that texture in view in the skybox is not negative-z, it's positive-z

If you flip-x then you get the same problem but in the x-direction, when the camera says it's looking at -x it will show +x

**Comment by mkvenkit** [comment]
Thanks for checking this. Are seeing the same problem in your code using the fix you suggested?

> For any of the sky box examples, a simple solution is to multiply the normal at texture lookup time by [-1, 1, 1]

**Comment by greggman** [comment]
No, like I mentioned above, my fix, using (-1, 1, 1) incorrectly swaps neg-x and pos-x. Maybe I'm confusing myself though. Both of them look correct. I'm going to have to think about it more

**Comment by mkvenkit** [comment]
I think the view you mentioned from my code is correct. As per the GL docs image below, and my cubemap, +Z is the buddha face. The +Z of the left-handed cubemap coordinate system should match with -Z in our right-handed coordinate system, and we are facing towards -Z. 

<img width="422" alt="image" src="https://github.com/webgpu/webgpu-samples/assets/35648873/523e2f2c-23e5-45d4-a1c6-4e7b16713840">

**Comment by greggman** [comment]
I get what you're saying but if put objects down each axis, with (-1, 1, 1) everything lines up. With (1, 1, -1) they don't

https://jsgist.org/?src=61b1cc5c2fdd879e517b7bcea4392abd

The whole thing is very confusing though.

**Comment by sciecode** [comment]
I might be able to provide some relevant pointers to the discussion.

There are, possibly, up to 3 different coordinate systems in play here. 
- Each cubemap has it's own coordinate system provided by the software origin / capture author. There is no guarantee that any one cubemap adheres to the same coordinates WebGPU expects.
- Despite not making it clear anywhere, WebGPU's texture-cube uses a Left-Handed, +X right, +Y up, +Z forward system. This is first point of contention, during texture upload, despite the naming convention used by the files, they might not represent the same orientations. (This, however, doesn't seem to be the issue here. According to cross visualization provided, the cubemap does indeed match WebGPU's coordinate system.)
- Finally, there's the user's coordinate system, which also doesn't need to be the same. So before sampling with `textureSample` the user must transform the coordinates they define to the one expected by WGSL (same as the one mentioned before).

If all of them are the same, there's no need to apply any scaling factor. It will simply work. Given that there's the need to, this tells us that they aren't the same. Which leads me to believe the user coordinate system is right-handed.

Now to the issue itself. There's no "correct" way to translate between left/right handedness.

You can either try to match what each axis represent (what greggman was doing by inverting x), but this leads invertion in one direction - in this case horizontal direction.

Another possibility is matching orientations. In the cubemap +Z represented forward orientation, but in user-land -Z represents forward orientation. So we simply replicate the original orientation on top of the one we use and have differing axis (what mkvenkit was doing by inverting z). But the direction matches the original.

It is more common to go with the approach of maintaining the original orientation, but I have seen both used. The advantage of maintaining orientation is that both cubemaps will look exactly the same even in systems with different notions of left/right handedness.

**Comment by kainino0x** [comment]
> * Despite not making it clear anywhere, WebGPU's texture-cube uses a Left-Handed, +X right, +Y up, +Z forward system. This is first point of contention, during texture upload, despite the naming convention used by the files, they might not represent the same orientations. (This, however, doesn't seem to be the issue here. According to cross visualization provided, the cubemap does indeed match WebGPU's coordinate system.)

I had to think about this for a ridiculously long time, but I think I convinced myself that you're correct. It's not a WebGPU thing - we just pass the coordinates through unchanged. We found that it was the same in all APIs. (WebGPU/D3D and WebGL/OpenGL have different framebuffer coordinate spaces, but that's cancelled out by the flip from NDC to FBC, so I think it doesn't end up being relevant.)

I initially thought cube textures didn't have a handedness - they just give you what you asked for: if you ask for the +X direction, you get a pixel from the +X side of the cube. But what gives them a handedness is which direction the pixels are arranged in UV space (== the order that pixels are uploaded in == the way they appear when you look at the image files in an image viewer) and that they're all centered around the +Z face, which results in what the OpenGL wiki says:

> For the top and bottom [and +x/-x] faces, it's like facing along the +Z axis and craning your head up and down [and right/left].

In other words they're left-handed when viewed from the inside, and right-handed when viewed from the outside.

Of course nothing in WebGPU forces the handedness of our world space, but we should follow common convention in this sample. So IMO the right place to fix this is by flipping the coordinate we're sampling, in Z.

**Comment by greggman** [comment]
I get what you’re saying but I feel like that ignored my example above where it all matches up. with. -z it does not match up.

Just flipping the -z and guessing it’s correct doesn’t seem like it’s enough. if you make a cube map of the real world then you want north, east, south, west in that cubemap to match north, east, south, west in your model of the same area. given the example I posted, -z doesn’t seem to do that. although their may be some other error

Maybe it would be good to expand this sample with a few objects to show the cube map aligns, similar to the example above

**Comment by kainino0x** [comment]
@greggman Sorry, I glossed over that comment because I thought what you said in the previous one agreed with what I was writing.

I don't think they should line up... When I load your jsgist, they do line up, but the image is also flipped - the text is backward. The text is correct with mult=(1,1,1) which seems to be because you have other flips going on in your fragment shader.

**Comment by kainino0x** [comment]
If we did want to make them line up, we would have to flip the projection matrix such that world space is left-handed. I tried hacking this by negating the aspect ratio parameter of `mat4.perspective()`, which does seem to work.

**Comment by kainino0x** [comment]
Pushed that hack up so you can see it:
https://github.com/webgpu/webgpu-samples/compare/main...kainino0x:webgpu-samples:main
https://kai.graphics/webgpu-samples/?sample=cubemap

**Comment by sciecode** [comment]
> what gives them a handedness is which direction the pixels are arranged in UV space (== the order that pixels are uploaded in == the way they appear when you look at the image files in an image viewer) and that they're all centered around the +Z face

@kainino0x Exactly, the cubemap handedness comes from the relation of how images tile with each other. But inside the API, it's related to how texture coordinates seamlessly transition to what is considered left/right/top/bottom of each face's UV space.

> Just flipping the -z and guessing it’s correct doesn’t seem like it’s enough. if you make a cube map of the real world then you want north, east, south, west in that cubemap to match north, east, south, west in your model of the same area. given the example I posted, -z doesn’t seem to do that. although their may be some other error

@greggman, Flipping X or Z isn't a general solution here. The important aspect is not exactly the scaling factor applied, but the overall meaning of what it's being done. If you have a different set of coordinates in either one of the 3 possible places, the solution will need to be adjusted accordingly. There isn't a general scaling factor that applies to any left/right hand transformation.

Coordinates systems are simply conventions used by one to represent orientation information. If I represent north with Y+ and someone else represents it with X-, the north hasn't changed, just how each person chose to represent it. That's exactly what I meant by matching orientation, we don't try to use a different coordinate system axis within ours, they have no equivalence. We try to match the original orientation, if the cubemap authors means that north, east, up are in a specific orientation, we maintain that same orientation, doesn't matter what their axis say.

I will validate what's happening in your linked example and respond shortly.

**Comment by sciecode** [comment]
I was able to take a look at the example, I don't understand some of the flipping that is occuring in the fragment shader. So I removed the direction flipping after the inverse proj/view calculation. Since the camera target is at the origin, we can simply using the world-position itself as the sampling direction.

[modified jsgist example](https://jsgist.org/?src=cb64%2CeJzdXAt328ax_itbJb2EbAmkKDtNGdu1Y1mpz4kfx7Kd5ko6NkgsKcQgwAuAlBRF__1-s-_Fg6Ls1Oe0bmMTuzuPnZ2ZnZ0d4Gori-Z8a7T1Cx__9PodO_p0-WN-wYKKl9X21s5WyasqyWbl1ujqemdrmqScxuPxeCvJYn4RnlXzFOPkw6Qsze_fyq1TCUGjDR0PapJnFc8qND-IkxWbpFFZPjwhoPRk69FJxtiDSZStovLRg7764TSyJMbgKl9grNP_oA9c-HfreqdGVbJniT5O5ou8qNiySIOzqlqUo37_nI9ni-V0mcWAzKooLcO8mKnmfsHLfFlMeKkadjG7Ms8I8_YPJxlNa4eN8_iSXRGj86iYJdmIDX5g8k__Div4PF9xVp1xFvNptEwrNUwNwZ87fQI-48nsrBqxvcHgrwIewPPokwQVlAQhEpZoWkQzroCvTzIlIsFGnJSLNLocsXGaTz4Bl4tJDYwmFUsTtEVylGHjPImrM4eLNmDBQ1KVjGQbJRkv2uYg-ArFWMGWi7llsMP3NOUXooV-7MZJwSdVkkOwkzxdzjPRNYsWI7a_oHGazKM7khBBAS3-NxBD50m2q4lRC8Y3lQX66-pKv888EznJ0DIt8jmWYhPNUfpSfroc51L_T7KTTCngT--eS1y9G5DtF_EiKqrL_nyZzJbJ7iC8COd5vEw5-O1hKmCq5KQem_Akldfo8rJK0E-sfXNOz_OoKhJMVDF5hed7O2zFJ_vXt-TWQeezq1DfYVCi89lZDe2s4LPZPMrCWVKdLcdhknusws7Lqr8XXvgT8AicZOBKKAubcehlVPHDaMKDMvmd77CrKX4_zdO82GEVv6icn9fbUnegAWWl9fwhi_PJkqYYTgoOXM9STk9BTw7okRMAjHgIhX4Dhmi57VL1vA5JpLpAoxo04-AGyncB5MNYI64uhGIfVZcpx1jDvtf7BtYRDHYY_i-nSX87CIAWsB-_vaJ2yH4Q_u16ccHKKCt3S14k049mLNF_kiazDAC9CWbKC5Kq7fwxKnkKg6f-eRLHKbf9LqdGul7vW5oedUkeWZ8N7S_JccGrZZEpqUhbhXMT8qLZH9Hgh2xv-D36bPNTMZzTih0TlisrqhHrfXM4GPScFaemweGhasLT3X_12PVOG-BhE3DgAO52AQ6agIcu4N1fOwH1qBbuiWInoEHfwj1R_N-uOTYBLfdEUQGehvNoERDo82yas4ePfAvTq7MjsNOQbVrRkywqL7MJM2Y5x34ReMYWxdECioa1i86jpGJZtEpmUZUXIYz8H2HB_2-J6OSJHBUoxRaQMV8lE9IHCagQGZAD0a0gkikL_iIBFHVsFFGSBr2M85i2wSI_hzlgn4sqVi4X5KlKtQkoe9TqKR6EZmJ77LOfOGjr7UJsIRdQS3Jt7oabxdQ3hR8vODbPdc4G3BeXRzyFYedF3dlIGEllr92BSP_oASwQypCrpjU4zAs4Z4B6kiYMrws-5UXBY2lQcqAVuSAZmlkESoxSqkK3IFMBM2ohqAZE6eIsepHHHLqFQXNERMkiTXjcEwOupdbYica0wB-ppayKJcKW90fvsoTIqHAHkUVepPH7hJ-_LvLfdKxA29fFvalEKmbweFbky0Uw2GaPx9j0EefS71VUPFhKhI_YqtS4Rw4dxZGi_-Ly_dHzbLHEFinJAzNCKDFTQrjIy0SygL1TMeAPGm6TcU3yvIjFoKHlskbn1bJyCD0eLxNIKws0iS5iltTeelKPV7xAv4iaYJug-ILsczWys9xmu49aeIHYIC20yaGyU1mJaA81a1g-K9awZa2wJ63MaA-FZp1QmAfPFOVA1yBbVhlCIHbjZDpdlvwomi9SDkdXyh9duoFVcqBo74LOj4RbxI8Pw_jBdH_4SAtyWkQzMl4jykNPlFI-QpaesohV0zJVU1IUJJ-BT3-nNgsEaEYw0k4hBsKlNnVhRAphDKtDqESeRlisCmqOzqKYF7JPmzSZXZs5ymBrUwwjabf0R-l03XLFUkMZDnSMb7XieQbdLIHDNWT5h3yG0OZ926o12iFWV9lbmVCbWddHuDQ3cC74MTISUFq8Tl9x9GzV1Q4opa-AaujqZDnmrrYKDNL0SVlXZWDkIls_iEMREL6McZRa7gM3NLdNoCk2P4iGNuGiiC4D24OVJX8T7O7tsP1tR6xOx-5eWwfbr3Xo7bfmeGpup9310IrRah_TXE53GLjZ8_BpCOGnXGhnUMPZiOV3hKnNn8Q5LYMagzfYvZYjTrpgAmoSrrcJcpiaU4fJPvIVkk0xZ2ALLy5_r0231cdYndlxtI58jUABesQUmZ1Fpf1M664NfcNMcMRTHuLH5RSBRfkki59UOBaOl0g4HSJCekI6UwYqghAAiyKZQ_wrXirYp0D2HlqJEWUggkonCvrzyLwtEKLi3xhhVI2epTjj-etkoY9AnhN8w2Eyhe7VbhD5jJw0oRctq1xGOKTlZGMju_zSrRp7VY7aGAD0qrh8nScZIdKbtEbG2FjOeiTPPvJPGIa0CKHs-1lwURqIU_VLHgiwYSnt_WyW5GZnWaqQ4uKVz9JVd2io-WhyZpbJYQ0pQJxUZkgU9bDIUTZL-W6K3IAlP1mmqQoxx9Hkk-7QSGO-qM6OkOmZJKmDVzT_UiQ452fROOWImLCR2DmL_qf5HEkOwkzpFEtST64nRg3vLdKl6ZV0azayuKUejTl47WU5i4xqa_y3UbL2FVzVeF2jEu0Ipo4o7OLjqNm55uy6vtZ_3rLs4gAYpZ-1OOagqldDZYBF3gKGYA6PchXV5i4dj8pOBHvfbcNl3lOusmVgfc1lq15ru9rCjypIR8CUMsGuXCdt-pclcsMjhuOo7HxHz-G7l88PX715wf6odzx99frXDwdHb7VAtChqrL-PUpyq6djIz9lhmkfV_lA41qAphD67p3cKFTiKbKDc3SyysFyOZdiA5BWkViOsgp5GuCmbrbjm0ewQ0QvFST0ymqiwskLyd03fAjmNju6GGCjO-omirsbq6Y7mAhIME6Ea6SDLx7SRWwa04TrbCh3DDULptxEveEaX8JpjFawhBER6ewcbvNRYWJDaGWqqwq6tu_WAERlZYC37jrFI3NmxyjSUOOhwhwxPi0-3ElVRiJGXx6DVY1dX7KoJVbK7nxaWpuIeBmvWTKGEtWhKVCEmrFm0mowXV6jgiEKJIwFsQgk3-3isECLIvIJqYZHIT4lZO3Pu8itr2KatDGleDbFILnj6RVxLDF_A1xx53AIJ5FQ4dMOLQqWHBVUytw6z4QroYBc6aFyUItm4oizjCq50b6DiPoeIgC5ybCn810DpBBME1UgRoDYYn2zEeJNpzW8iQ3BQTDLSdMuXNx0MADeW0g6b10ZK1v9F415E1Vn4-rnIzd_3Ruq5rp3S6t87pXIS4YgANo-JOfrvHjS8Pp2NJr7ZdIZfaz5C2uKv4efOqHsp9zZfSpNb4am5wiBvK1pw6wmvtYM9ogJyaDkyIlVwTN6aLn1Ot7EmCPZb_Bu2DVbiFgceet-44_VY6ax-A1YczG-LlVAS4jVYhxbr3uZYBbtrsCLt8BlYBbtrsN77PAkQu2uw3q9jvRHvJGCaXdI2D3dtyzjGmOF9UnT1Fw1W1IYtc8ARuJsWLObPpNWY18rSwo33elqWkKR623nB2Rhawy-ndeod8eymLC-oWrfm54hegd2eZpzLFOaGeTddvzdqVhKBGdf-yM5mZ1GafkDlySdeyX_iPrJHuyh8WMxw2cf-jXQyPvsqdGg-l19pPl-DDs3n9680nwYdHbTDepxYVjT6eYzNDmndR7TuA9qa45k5SuD6dX2u1DLacVRnd3FWNcf1Wx_W9Umvfkpff0b_khN6LY_0-adzJb98OkXZX4m0mrivRn49yZelxotngRgFa1OJFT9jyo1aFj5RlNW9BK8EAbBGxVcWav26ifl05wu0cm5E2qjyZozeVakIE5hR0V4nJ3vfUeoCQPv-umyeLtgoWaBTBTptuDZP0MwSfEGOYPMMwS3yA1cxTmpZKa7dRJqrhwosjUn6npqeFzIvipLRA15OimSBugWItiZDUGTjqEwmuqzCgmlJoogwL3A_EE3OyG16UrJeVjmWEXuwS5aBPBxVMKFSBIJm52c8Y-dcIbdAMJD4FeoSexM4WyfRRFeSecFFl_hlu_w5-0lQy6STDv0MxmSGlFgSyot1DAd--vTnDsYVN628q8Spt0gxn73N30R0YY__HuGvO-ZI1Gd73_seQFcb22WUN7t7ioI4VwkV0Xjp9KE86OIgP6daW_WMS4HDV--dkd8pa_AZJALOGeuYbgMp2ldYxAHIa5DPiNxb4rvVPjDBvh_haByGC1khJZMWYZUfIskS42CzHf6GJHmAzVVU5DgaHcVv8wM-e7WgSQop0G44ArXvYaTziKpY6VdZcUgfbABOpPXpsgdFpFSKo55DjUzlv1opDC0JIJP4__b3L8Cf0uVMg33N-99d1tHzBlc1pHyb07GUUP2q9jaMVxVJaAujOA60GkHAtLwQtFhlfzWO4BFxSb6tEkktsFrbAF-TWyeIVEIA7O7THbcYhjrW5kCpnU3MqupRzjERdXWazs_k0YKeX_Zlu2XSMSDMKBw7x6UvnIIa3L9zhz2uLhecXclh8k4JvKgSseHBtarY9kpRw-5i1EbpGelSV8HaN2Cqt722JE2g-DqlZI530ju_4wFQ_ofoV_s08XRhnkQE7j2pPu0H6NZep5ma5XlUBekmdJx8lSyu1DV1NuO9iCkxgHi-iPQZNZCrsk2a7GxQqniyXGBKtgxwguln1S-iFLnvN_5T1CE7pQC6MtDUYan82QJ2SFhx6-qVczD23aDuz73qDayIAKw1DkKYvN67fn-JDcbv36Okgu4-dHvdwgV0SrmwGdzpjALSSVJgv5KllnmR0EsVaZ5_ol7cL8qW-nQnAsdrWyByLOZTwkNre5X5PMo9wn_pNmnqKsE3ycuW0aQUPjGySS1V4uxJ5QnU58WTGiVRRCJuhyYub1Tr_SL5JfqXC0dm9qfWTM3CJF9cBsRTTa6_cJZn6SX4QbIiGlMgYUyFlbn7DouTB7UoCOXx3vBUhfh-83578z3d3L5ACNLpeBYIcUdVlA1hBq60jmnqtabh6Ta0UsKIHGx4f5uOd_TPHwxVN38VpzxLkTyuqGsPPp6cqOhQ6hj5nW-vVvs1qtuq6kUuiBilFsqDdl5Z-RaRrvY6x2Zup9cf7WWYry9OXaRaNuXXLgNrqij7qy2jl_n28cjBNxy1WmqErtrQyO2f_ujI1WTZutzb8D_FvRmTltv1n-Hb7m_k27pcE-UZcZIafG3PQtlQRLwtruW_Q2PdLdxcPMkzk39nSQ_sDhxVOBjsGYOl03nQ2OmRSPE2_sYF1ZXu3mkxBHZNWm7BDdMUYlzJg5jKGncOVacqSkn_iHSzGcQePjS_h-wfyFYjCnfcj3qHQbytsET5P8oYdeq6_hqDxojXGRxoyprQCxhIItEeoWBxOpUyxa_Q51Fi0_MxzImQUXKg-lSoL9e4mQDAluYf5rEtiOJJ4zIUHq82wZ_68ymdmGME_hU7i7DPwX-TtM08Xr2ht0fofUfxrlJSihpsTlw6aGqS0nMTJ3LxAmUkji8eag2uXk5xlviPP1ybdLvUe2Z_oSX1Zijb1wCqF9GakLLD0VfJkQvsddbwxnjXpsgvnaUy1aH1sfVcmF5lD7lMnWKzbMyOdvwm4zZVfVOtl_xjk68Kj0yyvnn28uDZmw9P3r598vSfL5691PVQpnRFPzimh7nXJmcf65WyHRrcnvDRSuwJukOHpUkBAU5CjWw1CuPwSmX8TPY6a6S2RzADEAUcjjlCZlWBiJ6gjWOLgoBD2L4pVtTJySZ37Z6ZsqoNT6gDAnvHjt2EyATiWH9sdml6PK0Z83JBOThhh7UEtkpv61aZ9tSQSmQ4xy6halRrqLL8fmESbYxeGrgpC5vdxViTA66Ni4voHJfTDusNUTolYe44Z-txbKZeA9VdPNVVatUoq0K-Q1-TuoPoStR5RqBE6qhiHJgu7T02t0lboixraDgXMZL9j9mtag7GDyPqymNYE7u1Ci0o60Rtrql-2cq6ePQCUWU4v1A4KNNF01Nl1rT7-NQ11HN6nUIBSQjxgoUm7jTJBEc7lpt0y9MvQRIJSIk7W87Vy8uOzjquyXmXQWCA3bf4mIl0JuZeTLuNaZIl5ZnjXDyp46oEacLg2IM2gjKU69GS8vjSx4uQB68RmxeD1VBqopS18xK0B-f4Owwtsb27LpDeajZVPnL_FO8kq53SvJWsBwv3-BrHy4YfRXoTHIqTJ35KXO08RMVEMUowVPKF3CGe7Km1RpQ210-S7RvmIuuQArzVXUtN4BCs0NdQi1Io_UbMxpKi44Fh3h_kvgQu32a-WYBoJzf3Nhdyae3YRdHF_dYe1tmzFll3zy6KDbsIretr72kyQRJqLqZYYys4enX7loLrFIJcKSyUfT0C_ZTRQT_VqKl-tntbHI7UupF0qC9yunSZVNdgp7VLrU62vplOpydbvnJ2fbnAHdD-9QIa1e9_9icDmqAbfzSgAbr5ZwNaQDf9cEATdONPB7SIabOPB3S7CGLyB_9bFsYx3oc24Xad9GsPXrlufC2-xv_URO8u0vWiwkSX1goMa5ghs_OZMY7udsxALnVm6O7A-s06M4hdx3wS4c1efdTGOfPur-KW5XwNyz4lyTLYNJO-Rxa5Ccv-3JX8cPkhSv4Mthvk186MmfTGzJASNeSnmLHY7P1Ol1cxwYX-igQ-QSPvkQqcONTxxnlv-qZhNEZu-PkYH4tAMkrdSqJMEKfWV6oxUAUYTk7TCdzFi1MUGqtRjUSR-SKEGBnKHF0t4NBfm5FD1IeL8KEiKvChJEiSkaejpxqg-RxNF6T4HpQP6H_gRkY50QUFDvWIRwV-4mKYbmAv1Nn1QJd7DA-cUv76F3JaMcve26GW1dXySa9UqH5458tmpwkyuwaYy0szSmiZySOKr4rMy5laVkRYBY6ueNYj5SdQBPat69Pr_we1BeEc)

This gives us a right-handed coordinate system (-Z forward, +Y up, +X right), you can validate this in the example by uncommenting the return with the direction itself.

That way, the only required change is inverting Z axis. And sure enough if you select the (1, 1, -1) multiplier in UI, the cubemaps appears the way I would expect them to.

Please note, the box in -Z/+Z positions won't say the same thing as the background behind it. Because we are not keeping the cubemap coordinate system intact, we are matching the orientation. The original orientation specified +Z is forward, and in your system -Z is forward. But if you construct the cross visualization of the cubemap, you'll note that everything looks indeed in the same orientation.

**Comment by sciecode** [comment]
A little while back I made two debug left/right handed cube textures left/right, to test exactly how to properly use them with WebGPU. If this is helpful in anyway, I wouldn't mind contributing these.

<img src="https://raw.githubusercontent.com/sciecode/phoptics.js/50f3e7151ef243a26ee59a64a29a879675c9ebf5/examples/textures/lh_cubemap/cubemap.jpg" width="400" height="300"> <img src="https://raw.githubusercontent.com/sciecode/phoptics.js/50f3e7151ef243a26ee59a64a29a879675c9ebf5/examples/textures/rh_cubemap/cubemap.jpg" width="400" height="300">

The only difference with the right-handed cubemap, is that during texture upload, the orientation-based flipping procedure needs to occur. In my example, the -Z texture is uploaded in 5th index (WebGPU posz) and +Z in the 6th index (WebGPU negz). Inside the shader, we still flip Z axis, for the reasons discussed previously.

**Comment by greggman** [comment]
Thanks all for the explanations. I finally got it. I agree flipping z is best.

I think it might be good to use a cubemap with some signs in it so it's easier to tell when something is flipped. I spent a few minutes looking but other than the shopping arcade from the example above I didn't find a good one. We could use that one. I just thought variety is good. A Taiwan night market, a Hong Kong sign filled street, a shopping mall. The hard part is getting a shot with no people or generating one by taking lots of shots. I don't have my 360 camera with me (and it's not that high-res).
